lib/renderer/src/renderer/scene/AbstractRenderer.h

```diff
@@ -0,0 +1,35 @@
+#ifndef VANGERS_ABSTRACTRENDERER_H
+#define VANGERS_ABSTRACTRENDERER_H
+
+#include "../common.h"
+#include "../ResourceId.h"
+
+namespace renderer::scene {
+	struct MapDescription {
+		int32_t width;
+		int32_t height;
+		uint8_t* material_begin_offsets;
+		uint8_t* material_end_offsets;
+		int32_t material_count;
+	};
+
+	enum class _HeightMapT;
+	typedef ResourceId<_HeightMapT> HeightMap;
+	static_assert (sizeof (HeightMap) == sizeof (int32_t), "invalid HeightMap RID size");
+
+
+	class AbstractRenderer {
+	public:
+		virtual HeightMap map_create(const MapDescription& map_description) = 0;
+		virtual void map_destroy(HeightMap map) = 0;
+		virtual void map_query(HeightMap map, int32_t* width, int32_t* height) = 0;
+		virtual void map_update_data(HeightMap map, const Rect& rect, uint8_t* height, uint8_t* meta) = 0;

```

Comment:
where is the data to update with?

Reply:
The idea is that we copy a slice of data from the `lineT` to the `height` and `meta` and pass them to the `map_update_data`. And delete them after the call. So may be the arguments should be `const uint8_t*`. I wrote some comments to the AbstractRenderer.h, see my last commits and [MapUpdater](https://github.com/KranX/Vangers/blob/d933a9e034892247a497a354e9c7cf21c2b7fdab/lib/renderer/src/renderer/scene/util/MapUpdater.h) class

Reply:
ok, yes, these should be `const uint8_t*` then